feat: ⚡️ minting metadata improvements (Metadatas + UI)

melMass opened the pull request:
UI part of #104 (mainly separated to get our staging link)

LeonNicholls committed:
Minting metadata improvements based on https://docs.google.com/document/d/1R3rxuJfpCBoh52wDsYpEmY_iuH8sd6l17oh4nk7jlRE/edit?usp=sharing

melMass commented on src/data/ipfs.js:
```diff
@@ -14,106 +16,77 @@ const infuraUrl = 'https://ipfs.infura.io:5001'
     tags,
     cid,
     address,
-    mimeType,
```
Is there a reason to remove mimetypes?

melMass replied:
Ah yes you resolve from extensions later on, I'm just refactoring this a bit

melMass committed:
feat: ✨ basic UI

Have to test it further but the logic seems to work.
UI still needs some polish.
This was minted through the ui:
https://ipfs.io/ipfs/Qmf6SwAejrJmYXCC5DQWvqZ2MXNqTdshbz94526MbGju8q

melMass committed:
fix: 🎨 unused imports

melMass commented:
I just went on integrating the UI and components.
Maybe we should use this PR to clean the mint file a bit

melMass committed:
feat: ⚡️ add lang select

I used Country codes (ISO) for now, but maybe we should use something
else.

melMass commented on src/constants.js:
```diff
@@ -166,3 +166,278 @@ export const MAIN_MARKETPLACE_CONTRACT = MARKETPLACE_CONTRACT_TEIA // the one th
+  custom: 'Custom (Specify)',
+}
+
+export const LANGUAGES = {
```
@NoRulesJustFeels Do you have a better source 😅 . 
I'm not sure Country ISO code is what was in the metadata proposal, but I improvised just to fill it for now

LeonNicholls replied:
This is the standard for the language based on the TZIP-21 spec: https://datatracker.ietf.org/doc/html/rfc1766
But it seems there have been other standards that have replaced this. Based on a search, I found this list: https://www.loc.gov/standards/iso639-2/php/code_list.php

melMass committed:
fix: 🔥 typo

melMass committed:
fix: ⚡️ use languages instead of countries

melMass commented:
Ok I'll leave it there for today, remaining:

- [x] Sort languages by language name rather than 'iso' code
- [x] Validate that rightsURI is a proper URL (must start with "https://" or "ipfs://"
- [x] Styling: Mixing the mint field animations with the new components or find a workaround that looks good (@SlinkousArt if you want to look into it too)
- [x] Clear fields does not work on "selects"
- [x] Product testers round

melMass committed:
fix: ⚡️ alphabetic order for languages

melMass committed:
feat: ⚡️ update select component

- adds a new dependency to the project:
  - react-select
  Selects are quite complicated when it comes to
  accessibility. This component is solving a lot of these
  common problems.

- Update the preview page to include the new informations:
  Adding it to the OBJKT UI should be done separately.
  NRJF opened #114 for that.

melMass committed:
fix: ⚡️ value was shadowing default value

This would close #115, but this turns all uncrontrolled inputs into
controlled ones.
https://reactjs.org/docs/uncontrolled-components.html
vs
https://reactjs.org/docs/forms.html#controlled-components

I think it should be done in a separate PR

melMass committed:
fix: 🎨 better wording of field names

melMass commented:
~~I should most probably revert b23cd16ad679bdad05fe7090baca10c3f6146ab5, just commenting to not forget.~~

Still needs testing but 26b9570bc81593e8c8cb4b8133afd22158c2de8e solves this

melMass committed:
fix: ⚡️ properly restore language and lic fields

melMass committed:
fix: 🎨 Unshadow other Input uses

Instead of reverting b23cd16ad679bdad05fe7090baca10c3f6146ab5
I unshadowed every Input usage (ie use defaultValue explicitely).

melMass committed:
fix: ⚡️ add checkbox component

melMass committed:
fix: ⚡️ reduce top padding

melMass committed:
fix: 🎨 style the select, fix input restore

melMass committed:
fix: 🎨 home search

melMass committed:
fix: ⚡️ text color on select (single value)

melMass committed:
fix: ⚡️ remove title from placeholder

melMass committed:
fix: ⚡️ remove margin on selects

melMass committed:
fix: ⚡️ hacky lines

@xat @SlinkousArt if you have a better approach I'm all ears :sweat_smile:

melMass committed:
fix: ⚡️ better checkboxes

melMass committed:
fix: ⚡️ add children to select

melMass committed:
fix: ⚡️ select's background color

melMass committed:
fix: ⚡️ match select labels with inputs

melMass committed:
fix: 🎨 margins

melMass committed:
Merge branch 'main' into mel/feature/metadata-ui

melMass committed:
fix: ⚡️ remove tag duplicates on blur

melMass committed:
fix: ⚡️ "validate" rightsUri

melMass committed:
fix: ⚡️ match collab inputs

IMO they should use the components, but this PR is starting to
get too big, especially since there is a refactoring going on.

melMass committed:
fix: ⚡️ remove hDAO from config

melMass committed:
fix: ⚡️ rightsURI can be undefined

melMass committed:
fix: ⚡️ Fix inputs on /config

melMass committed:
fix: ⚡️ actually validate the url

It's quite permissive, and not a full blown url checker.
"ipfs://" only would work for instance

melMass committed:
fix: 🎨 dens design feedback

melMass committed:
fix: ⚡️ remove console, and commented styles

xat commented:
the ui works nice 👍 

was wondering if we should maybe make the title mandatory? or are there cases where tokens are minted without title?

xat commented on src/components/input/index.js:
```diff
@@ -14,38 +14,70 @@ export const Input = ({
         max={max}
         maxLength={maxlength}
-        defaultValue={value}
+        defaultValue={defaultValue === null ? '' : defaultValue}
```
out of interest, why is defaultValue needed?

melMass replied:
This is one of the things I had the most issue with.
Basically it's the controlled vs uncontrolled input thing I linked in b23cd16ad679bdad05fe7090baca10c3f6146ab5 since now, the Input component was only exposing `value` but was used as `defaultValue`... So I unshadowed (26b9570bc81593e8c8cb4b8133afd22158c2de8e) that and kept using uncontrolled/controlled inputs where needed, it would have been easier to use a form library as you suggested 😅

xat commented on src/components/input/index.js:
```diff
@@ -14,38 +14,70 @@ export const Input = ({
         maxLength={maxlength}
-        defaultValue={value}
+        defaultValue={defaultValue === null ? '' : defaultValue}
+        value={value === null ? '' : value === undefined ? '' : value}
```
this might be easier to read:

```
value={(value === null || value === undefined) ? '' : value}

xat commented on src/components/preview/index.js:
```diff
@@ -28,9 +41,43 @@ export const Preview = ({ title, description, mimeType, previewUri, tags }) => {
+          {description}
+        </div>
+
+        <div className={styles.description}>
```
maybe rename `styles.description` to something more general, like `styles.field` (if it's not just used for the description)

xat commented on src/components/preview/index.js:
```diff
@@ -28,9 +41,43 @@ export const Preview = ({ title, description, mimeType, previewUri, tags }) => {
+            {photosensitiveSeizureWarning && (
+              <span
+                className={styles.label}
+                title={'This artwork is can cause seizures'}
```
the word `is` can be removed

xat commented on src/components/select/index.js:
```diff
@@ -0,0 +1,73 @@
+    ...provided,
+    backgroundColor: 'var(--background-color)',
+  }),
+  clearIndicator: (provided, state) => ({
```
can probably be removed, because no styles get changed

melMass replied:
Yes that's part of the leftovers I planned to remove, I'll do a pass with your edits and these later today 🚀

xat commented on src/constants.js:
```diff
@@ -166,3 +166,169 @@ export const MAIN_MARKETPLACE_CONTRACT = MARKETPLACE_CONTRACT_TEIA // the one th
+  custom: 'Custom (Specify below)',
+}
+
+export const LICENSE_TYPES_OPTIONS = Object.keys(LICENSE_TYPES).map((k) => {
```
afaik javascript does not guarantee a specific order of the object keys, but in practice, browsers order them as you would expect ('none' first).

melMass replied:
I did not know that 🙈 ! I guess we could ensure it with and `_.orderBy` but I tested a few browsers and it seemed consistent indeed (browser odering them)

xat replied:
yes, no need to change it, in practice it's not a problem. a Map guarantees the order btw, but that is clunky to use :)

melMass commented:
> was wondering if we should maybe make the title mandatory? or are there cases where tokens are minted without title?

Yes @merchantcoppola just wondered the same in #121, I have no opinion, but we could!

melMass committed:
fix: ⚡️ xat feedbacks

melMass committed:
Merge branch 'main' into mel/feature/metadata-ui

melMass commented:
I'm squash merging it, but won't delete the branch just yet just in case ;)
